### PR TITLE
Implement shellcheck

### PR DESCRIPTION
**Pull request by renekliment**
Our shell scripts also deserve some CI love.

### CODE REVIEW

**Comment by renekliment** [comment]
From Travis CI:

> ```
> $ cd src/scripts && shellcheck -e 2164 -x ./inc/*.sh ./*.sh
> shellcheck: unable to decommit memory: Invalid argument
> shellcheck: unable to decommit memory: Invalid argument
> ```

https://github.com/koalaman/shellcheck/issues/785

What to do about this? Merge it like it is and make sure we run tests locally to make sure it passes until there is an update and Travis can build this? I dunno :disappointed: I guess local tests are better than nothing.

**Comment by renekliment** [comment]
Okay, now it uses old shellcheck from trusty-backports, but it's better than nothing.

@alexa-pi/coreteam Please review.

**Comment by dgonano** [comment]
@renekliment i love your commit descriptions :rofl:

**Comment by renekliment** [comment]
Yeah, doing 13 commits just to try to make the damn thing install gets boring :smile:

**Comment by lechat** [comment]
@renekliment could you please rebase your fork and fixup/squash those commits?

File: src/scripts/inc/common.sh [hunk 1]
```diff
@@ -45,14 +45,15 @@ function init_systemd {
 function create_user {
 
     echo -n "Creating a user to run AlexaPi under ... "
-    UID_TEST=`id -u ${RUN_USER} >/dev/null 2>&1`
+    UID_TEST=$(id -u ${RUN_USER} >/dev/null 2>&1)
     UID_TEST="$?"
 
     if [ ${UID_TEST} -eq 0 ]; then
```
Comment (lechat) [inline src/scripts/inc/common.sh]:
```
if id -u ${RUN_USER} >/dev/null 2>&1; then
...
```
This shall work and does not create subshell and don't use special variables like `$?`

File: src/scripts/inc/common.sh [hunk 2]
```diff
@@ -45,14 +45,15 @@ function init_systemd {
...
         echo "user already exists. That's cool - using that."
     else
         useradd --system --user-group ${RUN_USER} 2>/dev/null
 
+        # shellcheck disable=SC2181
         if [ "$?" -eq "0" ]; then
```
Comment (lechat) [inline src/scripts/inc/common.sh]: Same as above.

**Comment by lechat** [comment]
While you are changing shell sctipt, I think that it will be good if we ask bash to be strict as well by adding these lines on top:

```
set -o nounset    # Fail when variable is used, but not initialized
set -o errexit       # Fail on inhandled error exits
set -o pipefail     # Fail when part of piped execution fails
```

I always put those when I expect my script to run with elevated permissions.

**Comment by renekliment** [comment]
@lechat Good points, I'll add it there.
Is there any advantage in squashing the commits now compared to doing that on the merge?

**Comment by lechat** [comment]
Well, when you do merge, you have two options: squash merge or merge with all commits. In my repos, I usually do merge commit, especially for PRs that have a lot of changes. That way if some merge introduces regression, I can find smaller change that made it. 

Also, when I make pull request, I usually do fixup/squash because it is much easier to review code that has:
a) meaningful commit messages and none of commits are fixes for typos and similar stuff
b) all commits that make actual changes

**Comment by renekliment** [comment]
Comments addressed. Thank you.

The strict mode should really help with buggy installs, because I think there were a couple of them when the install of some dependency (usually via pip) failed, but the user hasn't noticed and then filed an issue report that some module is missing or whatever.

By the way, using containers with a _temporary_ filesystem is awesome for testing installs and stuff. This is what I use: `sudo systemd-nspawn -D jessie-base -xb`, where the rootfs was created by `sudo debootstrap jessie jessie-base` into a btrfs subvolume.

**Comment by maso27** [comment]
For what it's worth, I just installed this branch and ran it successfully on a CHIP.

I think most of what you're doing is error-proofing, but it doesn't seem that the changes have hindered the installation process at all.

**Comment by renekliment** [comment]
@maso27 Great, thanks!

@lechat Do you think this is ready to go now?

**Comment by renekliment** [comment]
Rebased on latest `dev`.

@alexa-pi/team Could you someone give this a go please? (and maybe do a final test?) I believe there are no more issues with this. Thank you.

**Comment by dgonano** [comment]
```
pi@raspberrypi:/opt/AlexaPi $ sudo src/scripts/setup.sh 
TIP: When there is a value in brackets like [default_value], hit Enter to use it.
**Dion: I don't think this line should be here**

Which operating system are you using?
debian        - Debian, Raspbian, Armbian, Ubuntu or other Debian-based
archlinux     - Arch Linux or Arch Linux-based
Your OS [debian]: 
Which device are you using?
raspberrypi   - all Raspberry Pi variants
orangepi      - Orange Pi or another H3 based board
chip          - C.H.I.P.
magicmirror   - Integrating with Magic Mirror project (MMM-AlexaPi)
other         - other SBCs, desktops, or anything else
Your device [raspberrypi]: magicmirror
Do you want AlexaPi to run on boot?
You have these options: 
0 - NO
1 - yes, use systemd (default, RECOMMENDED and awesome)
2 - yes, use a classic init script (for a very old PC or an embedded system)
Which option do you prefer? [1]: 
Would you like to have AlexaPi restart when it crashes? (y/N)? y
Creating a user to run AlexaPi under ... user already exists. That's cool - using that.
gpio:x:997:pi,alexapi
audio:x:29:pi,pulse,alexapi,shairport-sync
Would you like to also install Airplay support (Y/n)? y
Get:1 http://mirrordirector.raspbian.org jessie InRelease [14.9 kB]             
Get:2 https://deb.nodesource.com jessie InRelease [3,914 B]                            
Hit http://archive.raspberrypi.org jessie InRelease                                           
Get:3 https://deb.nodesource.com jessie/main Sources [762 B]                           
Get:4 http://mirrordirector.raspbian.org jessie/main armhf Packages [8,982 kB]                                 
Get:5 https://deb.nodesource.com jessie/main armhf Packages [962 B]                        
Hit http://archive.raspberrypi.org jessie/main armhf Packages                              
Get:6 https://deb.nodesource.com jessie/main Translation-en_GB [162 B]
Get:7 https://deb.nodesource.com jessie/main Translation-en [162 B]                                       
Hit http://archive.raspberrypi.org jessie/ui armhf Packages                                 
Get:8 https://deb.nodesource.com jessie/main Translation-en_GB [162 B]                                   
Get:9 https://deb.nodesource.com jessie/main Translation-en [162 B]                                                                                
Get:10 https://deb.nodesource.com jessie/main Translation-en_GB [162 B]                      
Get:11 https://deb.nodesource.com jessie/main Translation-en [162 B]                                                               
Get:12 https://deb.nodesource.com jessie/main Translation-en_GB [162 B]                                                            
Get:13 https://deb.nodesource.com jessie/main Translation-en [162 B]                                                                                                                
Get:14 https://deb.nodesource.com jessie/main Translation-en_GB [162 B]                                                                                                             
Ign https://deb.nodesource.com jessie/main Translation-en_GB                                                                                                                        
Get:15 https://deb.nodesource.com jessie/main Translation-en [162 B]                                                                                                                
Ign https://deb.nodesource.com jessie/main Translation-en                                                                                                                           
Get:16 http://mirrordirector.raspbian.org jessie/contrib armhf Packages [37.5 kB]                                                                                                   
Get:17 http://mirrordirector.raspbian.org jessie/non-free armhf Packages [70.3 kB]                                                                                                  
Get:18 http://mirrordirector.raspbian.org jessie/rpi armhf Packages [1,356 B]                                                                                                       
Ign http://archive.raspberrypi.org jessie/main Translation-en_GB                                                                                                                    
Ign http://archive.raspberrypi.org jessie/main Translation-en                                                                                                                       
Ign http://archive.raspberrypi.org jessie/ui Translation-en_GB                                                                                                                      
Ign http://archive.raspberrypi.org jessie/ui Translation-en                                                                                                                         
Ign http://mirrordirector.raspbian.org jessie/contrib Translation-en_GB
Ign http://mirrordirector.raspbian.org jessie/contrib Translation-en
Ign http://mirrordirector.raspbian.org jessie/main Translation-en_GB
Ign http://mirrordirector.raspbian.org jessie/main Translation-en
Ign http://mirrordirector.raspbian.org jessie/non-free Translation-en_GB
Ign http://mirrordirector.raspbian.org jessie/non-free Translation-en
Ign http://mirrordirector.raspbian.org jessie/rpi Translation-en_GB
Ign http://mirrordirector.raspbian.org jessie/rpi Translation-en
Fetched 9,111 kB in 30s (300 kB/s)            
Reading package lists... Done
Reading package lists... Done
Building dependency tree        
Reading state information... Done
git is already the newest version.
libpulse-dev is already the newest version.
memcached is already the newest version.
python-dev is already the newest version.
python-pip is already the newest version.
python-setuptools is already the newest version.
swig is already the newest version.
vlc-nox is already the newest version.
libasound2-dev is already the newest version.
The following packages were automatically installed and are no longer required:
  libc-ares2 libffi5 libv8-3.14.5
Use 'apt-get autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 95 not upgraded.
**Dion: System exited unexpectedly** 
pi@raspberrypi:/opt/AlexaPi $ 
```

**Comment by dgonano** [comment]
Fails during this part: 
```
cd "${ALEXASRC_DIRECTORY}"

NEWPIP=$(run_pip --version | grep "pip 1.5")
NEWPIP=$?
if [ ${NEWPIP} -eq 1 ]; then
    run_pip install --no-cache-dir -r ./requirements.txt
else
    run_pip install -r ./requirements.txt
fi
```

**Comment by renekliment** [comment]
It works fine on my LXC/Jessie. Where do you test it? 

It probably exists because some command gives a non-zero exit status and we've set that option in bash (we don't want to overlook errors). 

I've cleaned up the code a bit. Could you find out now what exactly fails please? Thank you.

**Comment by lechat** [comment]
@renekliment the code block above will fail when `errexit` is set because exit code is not checked directly, so bash assumes that it is not checked.

Code that actually checks exit code:
```
if  run_pip --version | grep "pip 1.5"; then
    run_pip install --no-cache-dir -r ./requirements.txt
else
    run_pip install -r ./requirements.txt
fi
```

**Comment by renekliment** [comment]
@lechat I see, now I understand. Nevertheless, I already did that change just do clean up :smile: (with swapped if/else of course)

**Comment by dgonano** [comment]
yep, basic setup runs fine for me now

**Comment by renekliment** [comment]
Alright, great. Let me guys know when you think this should be merged.

**Comment by renekliment** [comment]
@dgonano Wait, is the "thumbs up" an approval? :rofl: If so, there are no notifications for it, so let's please not use it for that.

I actually wanted to ask you guys @alexa-pi/team how we should say that, because sometimes it's clear for the reviewer, but not the other person. Can we agree on an obvious comment like "Agree to merge." or the _Review approval_ button?

**Comment by lechat** [comment]
Agree to merge

**Comment by renekliment** [comment]
Dope. Merging the sh*t out if this.

**Pull request merged**